### PR TITLE
Kvc l243 bio trade input

### PR DESCRIPTION
**Pull request by kvcalvin**
#496

### CODE REVIEW

**Comment by bpbond** [comment]
@rynacui are you available to review?

File: R/module-helpers.R [hunk 1]
```diff
@@ -277,3 +277,39 @@ replace_GLU <- function(d, map, GLU_pattern = "^GLU[0-9]{3}$") {
...
+  } else if(income_group == "high") {
+    pcGDP_2010 %>%
+      filter(value > aglu.HIGH_GROWTH_PCGDP) ->
+      regions
+  } else if(income_group == "medium") {
+    pcGDP_2010 %>%
+      filter(value < aglu.HIGH_GROWTH_PCGDP, value > aglu.LOW_GROWTH_PCGDP) ->
+      regions
+  } else{
+    # ERROR! -- what do I do here?
```
Comment (bpbond) [inline R/module-helpers.R]: `stop("Unknown income_group")`

Comment (kvcalvin) [inline R/module-helpers.R]: At least I didn't have it print `bleeeee`

**Comment by rynacui** [comment]
@bpbond I'm on travel to NYC. Will review it on my train back this evening.

File: R/zchunk_L243.bio_trade_input.R [hunk 2]
```diff
@@ -58,266 +60,404 @@ module_aglu_L243.bio_trade_input_DISABLED <- function(command, ...) {
...
+      repeat_add_columns(tibble::tibble(year = MODEL_YEARS)) %>%
+      mutate(minicam.energy.input = BIOMASS.NAME) ->
+      L243.DeleteInput_RegBio
+
+    # Set up all of the new supply sectors.
+    # Copy sector information to each region.
+    # Note that "traded biomass" only goes in the BIOMASS.TRADE.REGION
+    A_bio_supplysector %>%
+      repeat_add_columns(tibble::tibble(region = GCAM_region_names$region)) %>%
+      mutate(region = if_else(traded == 1, BIOMASS.TRADE.REGION, region)) %>%
```
Comment (rynacui) [inline R/zchunk_L243.bio_trade_input.R]: how about `mutate(region = replace(region, traded == 1, BIOMASS.TRADE.REGION))`? 
If I remember correctly, `replace` is faster than `mutate + if_else`

File: R/zchunk_L243.bio_trade_input.R [hunk 3]
```diff
@@ -58,266 +60,404 @@ module_aglu_L243.bio_trade_input_DISABLED <- function(command, ...) {
...
+      mutate(technology = subsector.name) %>%
+      repeat_add_columns(tibble::tibble(year = MODEL_YEARS)) %>%
+      mutate(minicam.energy.input = if_else(subsector.name == DOMESTIC.BIOMASS.NAME, BIOMASS.NAME, TRADED.BIOMASS.NAME)) %>%
+      mutate(coefficient = 1) ->
+      L243.GlobalTechCoef_TotBio
+
+    # Add share-weights to the global technologies
+    L243.GlobalTechCoef_TotBio %>%
+      select(sector.name, subsector.name, technology, year) %>%
+      left_join_error_no_match(select(A_bio_subsector, subsector, share.weight), by=c("subsector.name" = "subsector")) ->
```
Comment (rynacui) [inline R/zchunk_L243.bio_trade_input.R]: space `by = c`

File: R/zchunk_L243.bio_trade_input.R [hunk 4]
```diff
@@ -58,266 +60,404 @@ module_aglu_L243.bio_trade_input_DISABLED <- function(command, ...) {
     L120.LC_bm2_R_LT_Yh_GLU <- get_data(all_data, "L120.LC_bm2_R_LT_Yh_GLU")
     L102.pcgdp_thous90USD_Scen_R_Y <- get_data(all_data, "L102.pcgdp_thous90USD_Scen_R_Y")
```
Comment (rynacui) [inline R/zchunk_L243.bio_trade_input.R]: silence package?

File: R/zchunk_L243.bio_trade_input.R [hunk 5]
```diff
@@ -58,266 +60,404 @@ module_aglu_L243.bio_trade_input_DISABLED <- function(command, ...) {
...
+        select(-MORE) ->
+        L243.SectorUseTrialMarket_Bio
+    }
+
+    # Set up all of the new subsectors.
+    # Copy subsector information to each region.
+    # Note that "traded biomass" only goes in the BIOMASS.TRADE.REGION, but the subsector name includes the original region name
+    A_bio_subsector_logit %>%
+      repeat_add_columns(tibble::tibble(region = GCAM_region_names$region)) %>%
+      mutate(subsector = if_else(traded == 1, paste(region, subsector), subsector)) %>%
```
Comment (rynacui) [inline R/zchunk_L243.bio_trade_input.R]: Again, may consider `mutate(subsector = replace(subsector, traded == 1, paste(region[traded == 1], subsector[traded == 1]) %>% mutate(region = replace(region, traded == 1, BIOMASS.TRADE.REGION))`

File: R/zchunk_L243.bio_trade_input.R [hunk 6]
```diff
@@ -58,266 +60,404 @@ module_aglu_L243.bio_trade_input_DISABLED <- function(command, ...) {
...
+      mutate(region = if_else(traded == 1, BIOMASS.TRADE.REGION, region)) %>%
+      mutate(logit.year.fillout = min(MODEL_YEARS)) %>%
+      select(-traded) ->
+      L243.SubsectorLogit_Bio
+
+    # Share weights for the regional biomass subsectors
+    L243.StubTech_TotBio %>%
+      select(region, supplysector, subsector) %>%
+      mutate(year.fillout = min(MODEL_YEARS)) %>%
+      left_join_error_no_match(select(A_bio_subsector, subsector, share.weight), by="subsector") ->
```
Comment (rynacui) [inline R/zchunk_L243.bio_trade_input.R]: space `by = `

File: R/zchunk_L243.bio_trade_input.R [hunk 7]
```diff
@@ -58,266 +60,404 @@ module_aglu_L243.bio_trade_input_DISABLED <- function(command, ...) {
...
+      mutate(technology = subsector, share.weight = 1) %>%
+      select(-year.fillout) %>%
+      repeat_add_columns(tibble::tibble(year = MODEL_YEARS)) ->
+      L243.TechShrwt_TradedBio
+
+    # Set share weights for total biomass in SSP4 low income regions.
+    # SSP4 assumes limited trade with low income regions; so, share weights are set to 0.1
+    # See Calvin et al. (2017) for documentation. https://doi.org/10.1016/j.gloenvcha.2016.06.010
+    L243.SubsectorShrwtFllt_TotBio %>%
+      filter(region %in% get_ssp4_regions(L102.pcgdp_thous90USD_Scen_R_Y, GCAM_region_names, "low"),
```
Comment (rynacui) [inline R/zchunk_L243.bio_trade_input.R]: nice function!

File: R/zchunk_L243.bio_trade_input.R [hunk 8]
```diff
@@ -58,266 +60,404 @@ module_aglu_L243.bio_trade_input_DISABLED <- function(command, ...) {
...
+    # Compute share weights for subsectors based on cropland area. Largest region gets shareweight of 1.
+    # This will mean that international biomass supply is roughly proportional to
+    # cropland area (larger regions produce more biomass).
+    L120.LC_bm2_R_LT_Yh_GLU %>%
+      filter(Land_Type == "Cropland", year == max(HISTORICAL_YEARS)) %>%
+      group_by(GCAM_region_ID) %>%
+      summarize(Cropland = sum(value)) %>%
+      ungroup() %>%
+      mutate(share.weight = Cropland / max(Cropland)) %>%
+      full_join(GCAM_region_names, by="GCAM_region_ID") %>%       # Using full join to ensure regions without cropland show up
```
Comment (rynacui) [inline R/zchunk_L243.bio_trade_input.R]: space `by = `

**Pull request merged**